### PR TITLE
Gateway Metrics in Dashboard

### LINKED ISSUE #4601: Support Gateways in the dashboard
## Feature Request

Currently, There is no way to get gateways information in the Linkerd2 dashboard. As this information is exposed through the `public-api`, It should not be hard to have a `Gateways` tab where we can show the same information from the CLI using a `MetricsTable`

### What problem are you trying to solve?

Remote gateways information in the dashboard

### How should the problem be solved?

Add a `Gateways` Tab into the Linkerd2 dashboard?

### Any alternatives you've considered?

There is the CLI already for this information, but as this info is fetched from public-api, It should be a pretty straight forward PR.

### How would users interact with this feature?

If you can, explain how users will be able to use this. Maybe some sample CLI
output?

### PR DESCRIPTION
**Pull request by tharun208**
Added Gateway Component in Linkerd Dashboard
Updated MetricsTable to support gateway metrics

Fixes #4601

### CODE REVIEW

**Comment by zaharidichev** [comment]
@tharun208 This is great. I wonder though whether a lot of this will change significantly once the changed @adleong has planned land ?

File: controller/api/util/api_utils.go [hunk 1]
```diff
@@ -116,6 +117,14 @@ type TapRequestParams struct {
 	LabelSelector string
 }
 
+// GatewayRequestParams conttains parameters that are used to build a
+// GatewayRequest
+type GatewayRequestParams struct {
+	RemoteClusterName string
```
Comment (adleong) [inline controller/api/util/api_utils.go:123]: is this unused?

File: controller/api/util/api_utils.go [hunk 2]
```diff
@@ -268,6 +277,28 @@ func BuildEdgesRequest(p EdgesRequestParams) (*pb.EdgesRequest, error) {
...
+	window := defaultMetricTimeWindow
+	if p.TimeWindow != "" {
+		_, err := time.ParseDuration(p.TimeWindow)
+		if err != nil {
+			return nil, err
+		}
+		window = p.TimeWindow
+	}
+	multiClusterNameSpace := p.GatewayNamespace
+	if multiClusterNameSpace == "" {
```
Comment (adleong) [inline controller/api/util/api_utils.go:292]: if unspecified, I think we want to leave this blank which should return gateways in all namespaces

File: web/srv/api_handlers.go [hunk 3]
```diff
@@ -438,3 +438,22 @@ func (h *handler) handleAPIResourceDefinition(w http.ResponseWriter, req *http.R
...
 	w.Write(resourceDefinition)
 }
+
+func (h *handler) handleAPIGateways(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	requestParams := util.GatewayRequestParams{
+		TimeWindow:        req.FormValue("window"),
+		GatewayNamespace:  req.FormValue("gatewayNamespace"),
+		RemoteClusterName: req.FormValue("remoteClusterName"),
+	}
+	gatewayRequest, err := util.BuildGatewayRequest(requestParams)
```
Comment (adleong) [inline web/srv/api_handlers.go:448]: I don't think we really need a brand new struct and helper method for building the api request.  The only properties we need are the time window and (optionally) the gateway namespace so we should be able to construct the request inline here easily enough.

File: controller/api/util/api_utils.go (hunk 1 above)
Comment (tharun208) [reply controller/api/util/api_utils.go:123]: it is used

File: web/srv/api_handlers.go (hunk 3 above)
Comment (tharun208) [reply web/srv/api_handlers.go:448]: to make it consistent with the code with other handler's. I added this, but one more reason for adding this is that in future if we try to get a particular gateway from a namespace

Comment (adleong) [reply web/srv/api_handlers.go:448]: I'm a big fan consistency but I think in this case it's more valuable to simplify and to reduce boilerplate.  If we need more functionality in the future we can always add it at that time.

Comment (tharun208) [reply web/srv/api_handlers.go:448]: sure @adleong will do the changes

File: controller/api/util/api_utils.go (hunk 1 above)
Comment (tharun208) [reply controller/api/util/api_utils.go:123]: it is used in `handleAPIGateways` handler.

Comment (adleong) [reply controller/api/util/api_utils.go:123]: can you show me exactly where in the `handleAPIGateways` handler?  I see it being set on line 446 but as far as I can tell, the `requestParams` object is only ever passed to BuildGatewayRequest which does not use the RemoteClusterName field.

Comment (tharun208) [reply controller/api/util/api_utils.go:123]: fixed it

File: web/srv/api_handlers.go (hunk 3 above)
Comment (tharun208) [reply web/srv/api_handlers.go:448]: > sure @adleong will do the changes, but I am wondering the `timeWindow` validation should be done at the helper method. So, I don't want to pollute the `handler` by putting all the logic in a single place.
> 
> And if I add the `timeWindow` validation at the handler level, then I need to introduce one more constant `defaultTimeWindow`.And, I don't want this @adleong

Comment (adleong) [reply web/srv/api_handlers.go:448]: I don't think the validation is complex enough to require a separate function.  This code is more readable to me to have the time window validation and defaulting in-line here in the handler.

Comment (tharun208) [reply web/srv/api_handlers.go:448]: @adleong, can I remove both the `BuildGatewayRequest` for now ?

Comment (adleong) [reply web/srv/api_handlers.go:448]: yes, that's what I'm suggesting.

Comment (tharun208) [reply web/srv/api_handlers.go]: @adleong removed it

**Comment by tharun208** [comment]
@adleong , are there any other changes need to be done?

File: web/srv/api_handlers.go [hunk 4]
```diff
@@ -438,3 +438,31 @@ func (h *handler) handleAPIResourceDefinition(w http.ResponseWriter, req *http.R
 	w.Header().Set("Content-Type", "text/yaml")
 	w.Write(resourceDefinition)
 }
+
+func (h *handler) handleAPIGateways(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	requestParams := util.GatewayRequestParams{
```
Comment (adleong) [inline web/srv/api_handlers.go:443]: Thanks for updating and removing the helper function!  However, I was getting at is that we can remove this params structure entirely and just construct the request directly.  Sorry if that wasn't clear.  What I had in mind was something like this:

```
	window := req.FormValue("window")
        if window == "" {
                window = "1m"
        }
	_, err := time.ParseDuration(window)
	if err != nil {
 		renderJSONError(w, err, http.StatusInternalServerError)
		return
	}
	gatewayRequest := &pb.GatewaysRequest{
		TimeWindow:        window,
		GatewayNamespace:  req.FormValue("gatewayNamespace"),
		RemoteClusterName: rreq.FormValue("remoteClusterName"),
	}
```

**Pull request merged**